README
JSON::Tiny

Minimal JSON with no dependencies.

DESCRIPTION

Lightweight, fast, pure-Perl JSON in a stand-alone module with only core
dependencies.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

DEPENDENCIES

Minimum Perl version: 5.8.4. See INCOMPATIBILITIES section in JSON::Tiny's POD.

SUPPORT AND DOCUMENTATION

Once installed, you may find documentation with the perldoc command.

    perldoc JSON::Tiny

You may also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=JSON-Tiny

    Search CPAN
        http://search.cpan.org/dist/JSON-Tiny/

    See the module's POD for additional info.

LICENSE AND COPYRIGHT

Copyright (C)2012-2014 David Oswald

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

See http://www.perlfoundation.org/artistic_license_2_0 for more information.


Changes
Revision history for Perl module JSON::Tiny

0.55 2016-04-25
  - Canonical object encoding.

0.54 2015-10-27
  - Fixed loss of large integer precision.

0.53 2015-01-25
  - Minor POD and code tweaks.

0.52 2015-01-25
  - Remove deprecated object-oriented API.

0.51 2015-01-25
  - Document '/' escaping.
  - Add from_json and to_json functions.
  - Deprecate Object-Oriented API.
  - Improved Boolean tests.

0.50 2014-08-05
  - Whitespace parsing simplification/optimization.

0.49 2014-05-18
  - Fix regression: $j->encode({a=>undef}) threw exception.

0.48 2014-05-17
  - Mini-optimization in number detection code.

0.47 2014-05-13
  - Number detection heuristics better match user expectations.

0.46 2014-03-06
  - POD revisions.
  - Tighten 'examples/', &error, tests.

0.45 2014-03-05
  - Streamline POD.

0.44 2014-03-05
  - Established RFC7159 compliance.
  - De-deprecate &j: document limitations.

0.43 2014-03-03
  - Slim the dist.
  - Deprecate &j: Ambiguities with RFC7159.

0.42 2014-02-20
  - Document die on failure for &j.
  - More tests.
  - u007f isn't mentioned in RFC4627.
  - Adapt Mojo::JSON updates.
  - Eradicate //.
  - Cleaner examples.

0.41 2014-02-19
  - ADD decode_json and encode_json functions.
  - Fix decoding error.
  - Handle encoding errors better.
  - Fix line numbers in error messages.
  - Test all decoding errors.

0.40 2014-01-16
  - Mini-optimization: &encode.
  - Linkify RFC mention in POD.
  - Fixed bug with PREREQ_PM hashref.

0.39 2013-12-04
  - POD refinements.

0.38 2013-12-03
  - ACKNOWLEDGEMENTS recognize chansen's GitHub Gist: Mojo::JSON's birth.

0.37 2013-11-21
  - Removed minimum dependency version checks in Makefile.PL.
  - POD: JSON::Tiny is not relaxed.

0.36 2013-11-08
  - Special char \b (was \x07) correctly associated with 0x08.
  - Tests for 0x07=>0x08.

0.35 2013-10-16
  - References to a scalar (even blessed) encode as Boolean.

0.34 2013-10-15
  - Added t/22-bool.t to test Boolean override.
  - Documented Boolean override.

0.33 2013-10-15
  - Change 'my $TRUE', 'my $FALSE' to 'our': users can override
    Booleans. http://perlmonks.org/?node_id=1058232
  - Added META: Git repo, Meta spec versn.

0.32 2013-06-22
  - &j dies on non-parsable JSON.
  - Performance tweaks.
  - Document: Perl 5.8.9 or older may segfault decoding JSON strings >22k.

0.31 2013-06-19
  - Document Exporter dependency for Perl <5.8.4.
  
0.30 2013-06-19
  - Push minimum Perl version back to 5.8.4, was 5.10. (tye)
  - Document INCOMPATIBILITIES: upgrade Exporter to facilitate running under
    Perl <5.8.4.

0.29 2013-06-18
  - Added tests for empty keys.
  - Revised UTF patterns.
  
0.28 2013-05-31
  - Tests for "inf" and "nan" made portable. RT# 85775.
    
0.27 2013-05-15
  - Encode "inf" and "nan" values as strings.

0.26 2013-05-13
  - Change heuristics for number detection: better match user expectations.

0.25 2012-03-05
  - Add test "Decode object with duplicate keys".

0.24 2012-03-01
  - POD tweaks.

0.23 2013-02-27
  - Add &j, and Exporter dependency.
  - POD tweaks, incl. documentation for &j.
  - &j tests.
  
0.22 2012-11-02
  - Enable lexical warnings in Tiny.pm.
  - Silence unwanted Perl Critiques.
  - "Changes" W3CDTF compliance.

0.21 2012-10-27
  - Add boolean shortcut support to JSON::Tiny.
  - Fix context bugs.

0.20 2012-10-04
  - Bump to version number .20: avoid confusion with like-named Perl6 module.
  - POD tweaks.
  - Drop ref($class)||$class; from c'tor.
  
0.01 2012-10-03
  - JSON::Tiny adapts Mojo::JSON.
  - t/20-mojo-json.t adapts Mojolicious/t/mojo/json.t
  - Mojolicious dependencies purged.
  - Artistic 2.0 license, as Mojolicious.


examples/json_pp_benchmark.pl
#!/usr/bin/env perl

use v5.14;

BEGIN { $ENV{PERL_JSON_BACKEND} = 0; } # Force JSON::PP.

use strict;
use warnings;
use File::Slurp;
use JSON;
use JSON::Tiny;
use Benchmark qw( cmpthese );

my @json = do { split /-{5,}/, read_file(\*DATA); };

sub json_pp {
    my $j = JSON->new->relaxed;
    my @decoded = map { $j->decode( $_ ) } @json;
    return \@decoded;
}

sub json_tiny {
    my $j = JSON::Tiny->new;
    my @decoded = map { $j->decode( $_ ) } @json;
    return \@decoded;
}

cmpthese ( -5, {
    JSON_PP   => \&json_pp,
    JSON_Tiny => \&json_tiny,
} );



__DATA__
{
    "firstName": "John",
    "lastName": "Smith",
    "age": 25,
    "address": {
        "streetAddress": "21 2nd Street",
        "city": "New York",
        "state": "NY",
        "postalCode": "10021"
    },
    "phoneNumber": [
        {
            "type": "home",
            "number": "212 555-1234"
        },
        {
            "type": "fax",
            "number": "646 555-4567"
        }
    ]
}
------------------------------------
{
    "name": "Product",
    "properties": {
        "id": {
            "type": "number",
            "description": "Product identifier",
            "required": true
        },
        "name": {
            "type": "string",
            "description": "Name of the product",
            "required": true
        },
        "price": {
            "type": "number",
            "minimum": 0,
            "required": true
        },
        "tags": {
            "type": "array",
            "items": {
                "type": "string"
            }
        },
        "stock": {
            "type": "object",
            "properties": {
                "warehouse": {
                    "type": "number"
                },
                "retail": {
                    "type": "number"
                }
            }
        }
    }
}
------------------------------------------
{
    "id": 1,
    "name": "Foo",
    "price": 123,
    "tags": [ "Bar", "Eek" ],
    "stock": {
        "warehouse": 300,
        "retail": 20
    }
}
-------------------------------------------
{"web-app": {
  "servlet": [   
    {
      "servlet-name": "cofaxCDS",
      "servlet-class": "org.cofax.cds.CDSServlet",
      "init-param": {
        "configGlossary:installationAt": "Philadelphia, PA",
        "configGlossary:adminEmail": "ksm@pobox.com",
        "configGlossary:poweredBy": "Cofax",
        "configGlossary:poweredByIcon": "/images/cofax.gif",
        "configGlossary:staticPath": "/content/static",
        "templateProcessorClass": "org.cofax.WysiwygTemplate",
        "templateLoaderClass": "org.cofax.FilesTemplateLoader",
        "templatePath": "templates",
        "templateOverridePath": "",
        "defaultListTemplate": "listTemplate.htm",
        "defaultFileTemplate": "articleTemplate.htm",
        "useJSP": false,
        "jspListTemplate": "listTemplate.jsp",
        "jspFileTemplate": "articleTemplate.jsp",
        "cachePackageTagsTrack": 200,
        "cachePackageTagsStore": 200,
        "cachePackageTagsRefresh": 60,
        "cacheTemplatesTrack": 100,
        "cacheTemplatesStore": 50,
        "cacheTemplatesRefresh": 15,
        "cachePagesTrack": 200,
        "cachePagesStore": 100,
        "cachePagesRefresh": 10,
        "cachePagesDirtyRead": 10,
        "searchEngineListTemplate": "forSearchEnginesList.htm",
        "searchEngineFileTemplate": "forSearchEngines.htm",
        "searchEngineRobotsDb": "WEB-INF/robots.db",
        "useDataStore": true,
        "dataStoreClass": "org.cofax.SqlDataStore",
        "redirectionClass": "org.cofax.SqlRedirection",
        "dataStoreName": "cofax",
        "dataStoreDriver": "com.microsoft.jdbc.sqlserver.SQLServerDriver",
        "dataStoreUrl": "jdbc:microsoft:sqlserver://LOCALHOST:1433;DatabaseName=goon",
        "dataStoreUser": "sa",
        "dataStorePassword": "dataStoreTestQuery",
        "dataStoreTestQuery": "SET NOCOUNT ON;select test='test';",
        "dataStoreLogFile": "/usr/local/tomcat/logs/datastore.log",
        "dataStoreInitConns": 10,
        "dataStoreMaxConns": 100,
        "dataStoreConnUsageLimit": 100,
        "dataStoreLogLevel": "debug",
        "maxUrlLength": 500}},
    {
      "servlet-name": "cofaxEmail",
      "servlet-class": "org.cofax.cds.EmailServlet",
      "init-param": {
      "mailHost": "mail1",
      "mailHostOverride": "mail2"}},
    {
      "servlet-name": "cofaxAdmin",
      "servlet-class": "org.cofax.cds.AdminServlet"},
 
    {
      "servlet-name": "fileServlet",
      "servlet-class": "org.cofax.cds.FileServlet"},
    {
      "servlet-name": "cofaxTools",
      "servlet-class": "org.cofax.cms.CofaxToolsServlet",
      "init-param": {
        "templatePath": "toolstemplates/",
        "log": 1,
        "logLocation": "/usr/local/tomcat/logs/CofaxTools.log",
        "logMaxSize": "",
        "dataLog": 1,
        "dataLogLocation": "/usr/local/tomcat/logs/dataLog.log",
        "dataLogMaxSize": "",
        "removePageCache": "/content/admin/remove?cache=pages&id=",
        "removeTemplateCache": "/content/admin/remove?cache=templates&id=",
        "fileTransferFolder": "/usr/local/tomcat/webapps/content/fileTransferFolder",
        "lookInContext": 1,
        "adminGroupID": 4,
        "betaServer": true}}],
  "servlet-mapping": {
    "cofaxCDS": "/",
    "cofaxEmail": "/cofaxutil/aemail/*",
    "cofaxAdmin": "/admin/*",
    "fileServlet": "/static/*",
    "cofaxTools": "/tools/*"},
 
  "taglib": {
    "taglib-uri": "cofax.tld",
    "taglib-location": "/WEB-INF/tlds/cofax.tld"}}}
